#### Aquí encontrarás todo lo que necesitas saber: https://www.kaggle.com/t/ab8726f0cfc84544abbae69a6be88071

## Librerias

In [ ]:
#AHORA DEJO LAS COLUMNAS DE CPU Y RAM

In [1]:
import pandas as pd
import matplotlib.pyplot as plt
import numpy as np

import urllib.request
from PIL import Image

## Data

In [2]:
### Para que funcione necesitas bajarte los archivos de datos de Kaggle 
train = pd.read_csv("train.csv")


## Data exploration

In [3]:
train.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 912 entries, 0 to 911
Data columns (total 13 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   id                912 non-null    int64  
 1   Company           912 non-null    object 
 2   Product           912 non-null    object 
 3   TypeName          912 non-null    object 
 4   Inches            912 non-null    float64
 5   ScreenResolution  912 non-null    object 
 6   Cpu               912 non-null    object 
 7   Ram               912 non-null    object 
 8   Memory            912 non-null    object 
 9   Gpu               912 non-null    object 
 10  OpSys             912 non-null    object 
 11  Weight            912 non-null    object 
 12  Price_euros       912 non-null    float64
dtypes: float64(2), int64(1), object(10)
memory usage: 92.8+ KB


In [4]:
train.head()

id Company                              Product            TypeName  \
0   852    Dell                        Inspiron 5767            Notebook   
1   414    Asus                         ZenBook Flip  2 in 1 Convertible   
2   283  Lenovo                   Ideapad 320-15IKBN            Notebook   
3   669    Acer                             Aspire 3            Notebook   
4  1287      HP  15-AC110nv (i7-6500U/6GB/1TB/Radeon            Notebook   

   Inches                           ScreenResolution  \
0    17.3                          Full HD 1920x1080   
1    13.3  IPS Panel Full HD / Touchscreen 1920x1080   
2    15.6                          Full HD 1920x1080   
3    15.6                                   1366x768   
4    15.6                                   1366x768   

                          Cpu  Ram     Memory                    Gpu  \
0  Intel Core i7 7500U 2.7GHz  8GB    1TB HDD     AMD Radeon R7 M445   
1  Intel Core i5 7200U 2.5GHz  8GB  256GB SSD  Intel HD Graphics 620   
2  Intel Core i5 7200U 2.5GHz  6GB  256GB SSD  Intel HD Graphics 620   
3    Intel Core i3 6006U 2GHz  4GB  128GB SSD  Intel HD Graphics 520   
4  Intel Core i7 6500U 2.5GHz  6GB    1TB HDD     AMD Radeon R5 M330   

        OpSys  Weight  Price_euros  
0       Linux  2.83kg        865.0  
1  Windows 10  1.27kg        928.0  
2  Windows 10   2.2kg        579.0  
3       Linux   2.1kg        412.0  
4  Windows 10  2.19kg        764.0

In [5]:
train.corr()

id    Inches  Price_euros
id           1.000000 -0.087431     0.069513
Inches      -0.087431  1.000000     0.085601
Price_euros  0.069513  0.085601     1.000000

In [6]:
import seaborn as sns

# sns.pairplot(train)

In [7]:
# plt.figure(figsize=(10,5))

# sns.heatmap(train.corr(),
#            vmin=-0.1,
#            vmax=1,
#            cmap=sns.diverging_palette(145, 280, s=85, l=25, n=7),
#            annot=True);

## Data processing

In [8]:
# aqui tenemos que procesar la informacion

In [9]:
train.columns

Index(['id', 'Company', 'Product', 'TypeName', 'Inches', 'ScreenResolution',
       'Cpu', 'Ram', 'Memory', 'Gpu', 'OpSys', 'Weight', 'Price_euros'],
      dtype='object')

In [10]:
# IMPRIMO LOS VARLOES UNICOS DE CADA VARIABLE PARA VER QUE VOY A HACER CON ELLO

print('Compañia:',train['Company'].nunique(),'\n')
print('Product:',train['Product'].nunique(),'\n')# esta variable me la cargo o no la tengo en cuenta en el análisis
print('TypeName:',train['TypeName'].nunique(),'\n')
print('ScreenResolution:',train['ScreenResolution'].nunique(),'\n')
print('Cpu:',train['Cpu'].nunique(),'\n')
print('Ram:',train['Ram'].nunique(),'\n')
print('Gpu:',train['Gpu'].nunique(),'\n')
print('OpSys:',train['OpSys'].nunique(),'\n')# es el procesador
print('Weight:',train['Weight'].nunique(),'\n')#convertir a numero directamente


Compañia: 19 

Product: 472 

TypeName: 6 

ScreenResolution: 37 

Cpu: 99 

Ram: 9 

Gpu: 100 

OpSys: 8 

Weight: 167 



In [11]:
train['TypeName'].unique()

array(['Notebook', '2 in 1 Convertible', 'Gaming', 'Ultrabook', 'Netbook',
       'Workstation'], dtype=object)

In [12]:
#AQUI AGRUPO POR TYPO, SACO EL MAXIMO Y LO ORDENO  
# EN FUNCION DE ESTE ORDEN LE DOY LOS VALORES A TYPENAME: 
# ESTE DICCIONARIO ES EL QUE VOY A UTILIZAR PARA REEMPLAZAR LOS VALORES EN LA SIGUIENTE CELDA
dic_typename= {'Notebook': 5, '2 in 1 Convertible':2, 'Gaming' :6, 'Ultrabook' :3, 'Netbook' :1,'Workstation':4}

train.groupby(['TypeName'])[['Price_euros']].max().sort_values(by='Price_euros',ascending=False)

Price_euros
TypeName                       
Gaming                   6099.0
Notebook                 4899.0
Workstation              4389.0
Ultrabook                2858.0
2 in 1 Convertible       2824.0
Netbook                  1908.0

In [13]:
# AQUI REEMPLAZO LOS VALORES DE LA COLUMNA TYPENAME

train['TypeName']= train['TypeName'].replace(dic_typename)

In [14]:
train['Company'].unique()

array(['Dell', 'Asus', 'Lenovo', 'Acer', 'HP', 'MSI', 'Microsoft',
       'Toshiba', 'Fujitsu', 'Razer', 'Mediacom', 'Apple', 'Samsung',
       'Chuwi', 'LG', 'Huawei', 'Vero', 'Google', 'Xiaomi'], dtype=object)

In [15]:
#AQUI AGRUPO POR COMPANY, SACO EL MAXIMO Y LO ORDENO  
# EN FUNCION DE ESTE ORDEN LE DOY LOS VALORES A TYPENAME:
#ESTE DICCIONARIO ES EL QUE VOY A UTILIZAR PARA REEMPLAZAR LOS VALORES

dic_company = {'Dell':15, 'Asus':16, 'Lenovo':18, 'Acer':11, 'HP' :17, 'MSI':13, 'Microsoft':12,
       'Toshiba':10, 'Fujitsu':4, 'Razer' : 19, 'Mediacom':3, 'Apple':14, 'Samsung':8,
       'Chuwi':1, 'LG':9, 'Huawei':6, 'Vero':2, 'Google':7, 'Xiaomi':5}

train.groupby(['Company'])[['Price_euros']].max().sort_values(by='Price_euros',ascending=False)



Price_euros
Company               
Razer          6099.00
Lenovo         4899.00
HP             4389.00
Asus           3975.00
Dell           3659.40
Apple          2858.00
MSI            2729.00
Microsoft      2589.00
Acer           2299.00
Toshiba        2226.00
LG             2099.00
Samsung        1849.00
Google         1559.00
Huawei         1499.00
Xiaomi         1399.95
Fujitsu         799.00
Mediacom        389.00
Vero            260.00
Chuwi           248.90

In [16]:
#AQUI REEMPLAZO LOS VALORES DE COMPANY POR LOS VAMORES DEL DICCIONARIO

train['Company']= train['Company'].replace(dic_company)

In [17]:
train['Ram'].unique()

array(['8GB', '6GB', '4GB', '16GB', '24GB', '12GB', '64GB', '32GB', '2GB'],
      dtype=object)

In [18]:
# ESTE DICCIONARIO ES PARA CAMBIAR LOS VALORES DE LA RAM
# CAMBIO LOS VALORES DE RAM POR LOS DEL DICCIONARIO
dic_ram = {'8GB' : 4, '6GB': 3, '4GB' : 2, '16GB' : 6, '24GB' : 7, '12GB' : 5, '64GB': 9, '32GB': 8, '2GB' :1}
train ['Ram'] =train['Ram'].replace(dic_ram)



In [19]:
train['OpSys'].unique()

array(['Linux', 'Windows 10', 'Windows 10 S', 'No OS', 'Windows 7',
       'Chrome OS', 'macOS', 'Mac OS X'], dtype=object)

In [20]:

#AQUI AGRUPO POR OPSYS, SACO EL MAXIMO Y LO ORDENO  
# EN FUNCION DE ESTE ORDEN LE DOY LOS VALORES A TYPENAME:
#ESTE DICCIONARIO ES EL QUE VOY A UTILIZAR PARA REEMPLAZAR LOS VALORES

dic_OpSys = {'Linux':1, 'Windows 10': 8, 'Windows 10 S':5, 'No OS':2, 'Windows 7' :7,
       'Chrome OS':3, 'macOS':6, 'Mac OS X':4}

train.groupby(['OpSys'])[['Price_euros']].max().sort_values(by='Price_euros',ascending=False)

Price_euros
OpSys                    
Windows 10        6099.00
Windows 7         4389.00
macOS             2858.00
Windows 10 S      2589.00
Mac OS X          2139.97
Chrome OS         1559.00
No OS             1399.95
Linux             1099.00

In [21]:
#AQUI REEMPLAZO LOS VALORES DEL OPSYS POR LOS VALORE DEL DICCIONARIO

train ['OpSys'] =train['OpSys'].replace(dic_OpSys)

In [22]:

# AQUI ESTOY AGRUPANDO POR LA COLUMNA GPU, Y MOSTRANDO EL PRECIO MAXIMO POR GPU ORDENADOR DE FORMA DESCENDENTE Y SETEO EL INDICE DEL NUEVO DATAFRAME CREADO PARA PODER DESPUES ITERAR A TRAVES DE LOS INDICES


df_grou_gpu = train.groupby(['Gpu'])[['Price_euros']].max().sort_values(by='Price_euros',ascending=False)
df_grou_gpu.reset_index(inplace=True)

In [23]:

# AQUI ESTOY AGRUPANDO POR LA COLUMNA GPU, Y MOSTRANDO EL PRECIO MAXIMO POR GPU ORDENADOR DE FORMA DESCENDENTE Y SETEO EL INDICE DEL NUEVO DATAFRAME CREADO PARA PODER DESPUES ITERAR A TRAVES DE LOS INDICES


df_grou_gpu = train.groupby(['Gpu'])[['Price_euros']].max().sort_values(by='Price_euros',ascending=False)
df_grou_gpu.reset_index(inplace=True)
#ADEMAS LE ESTOY AÑADIENDO LA COLUMNA GPU_NORM QUE ES A LA QUE LE VOY A DAR EL VALOR QUE VA A IR DESDE 1 A 0 (POR QUÉ ESTE VALR, PORQUE HAY 100 FILAS EN EL DATAFRAME AGRUPADO, ES DECIR, 100 TIPOS DISNTINTOS DECPU), ASIGNANDO EL MAYOR VALOR AL PRECIO MAS CARO.
df_grou_gpu['Gpu_norm'] =0
df_grou_gpu
valor = 0.99
for i in range((df_grou_gpu.shape[0])):
    df_grou_gpu.iat[i,df_grou_gpu.columns.get_loc('Gpu_norm')]=valor
    valor -= 0.01
    
# DESPUES PASO A DICCIONARIO LA COLUMNA GPU Y GPU_NORM PARA DESPUES REEMPLAZARLO EN TRAIN
dic_gpu = pd.Series(df_grou_gpu.Gpu_norm.values, index=df_grou_gpu.Gpu).to_dict()

#AQUI HAGO EL REPLACE DE LOS VALORES DEL DICCIONARIO EN TRAIN
train['Gpu']= train['Gpu'].replace(dic_gpu)

In [24]:
#ADEMAS LE ESTOY AÑADIENDO LA COLUMNA GPU_NORM QUE ES A LA QUE LE VOY A DAR EL VALOR QUE VA A IR DESDE 1 A 0 (POR QUÉ ESTE VALR, PORQUE HAY 100 FILAS EN EL DATAFRAME AGRUPADO, ES DECIR, 100 TIPOS DISNTINTOS DECPU), ASIGNANDO EL MAYOR VALOR AL PRECIO MAS CARO.
df_grou_gpu['Gpu_norm'] =0
df_grou_gpu

Gpu  Price_euros  Gpu_norm
0   Nvidia GeForce GTX 1080       6099.0         0
1      Nvidia Quadro M2200M       4899.0         0
2      Nvidia Quadro M2000M       4389.0         0
3   Nvidia GeForce GTX 980        3975.0         0
4      Nvidia Quadro M3000M       3949.4         0
..                      ...          ...       ...
95          AMD R4 Graphics        399.0         0
96            AMD Radeon R4        398.0         0
97            AMD Radeon R3        379.0         0
98            AMD Radeon R2        349.0         0
99   AMD Radeon R2 Graphics        299.0         0

[100 rows x 3 columns]

In [25]:
df_grou_gpu.shape[0]

100

In [26]:
# AQUI ESTOY ITRENANDO A TRAVES DEL DATAFRAME CREADO PARA AÑADIR EL VALOR CORREPONDEINETE A LA COLUMNA GPU_NORM
valor = 0.99
for i in range((df_grou_gpu.shape[0])):
    df_grou_gpu.iat[i,df_grou_gpu.columns.get_loc('Gpu_norm')]=valor
    valor -= 0.01

In [27]:

# DESPUES PASO A DICCIONARIO LA COLUMNA GPU Y GPU_NORM PARA DESPUES REEMPLAZARLO EN TRAIN
dic_gpu = pd.Series(df_grou_gpu.Gpu_norm.values, index=df_grou_gpu.Gpu).to_dict()

In [28]:
#AQUI HAGO EL REPLACE DE LOS VALORES DEL DICCIONARIO EN TRAIN
train['Gpu']= train['Gpu'].replace(dic_gpu)

In [29]:

# AQUI ESTOY AGRUPANDO POR LA COLUMNA CPU, Y MOSTRANDO EL PRECIO MAXIMO POR CPU ORDENADOR DE FORMA DESCENDENTE Y SETEO EL INDICE DEL NUEVO DATAFRAME CREADO PARA PODER DESPUES ITERAR A TRAVES DE LOS INDICES


df_grou_cpu = train.groupby(['Cpu'])[['Price_euros']].max().sort_values(by='Price_euros',ascending=False)
df_grou_cpu.reset_index(inplace=True)
df_grou_cpu

Cpu  Price_euros
0            Intel Core i7 7820HK 2.9GHz       6099.0
1          Intel Xeon E3-1535M v6 3.1GHz       4899.0
2          Intel Xeon E3-1535M v5 2.9GHz       4389.0
3            Intel Core i7 6820HK 2.7GHz       3975.0
4            Intel Core i7 6700HQ 2.6GHz       3949.4
..                                   ...          ...
94              AMD E-Series 7110 1.8GHz        297.0
95              Intel Atom Z8350 1.92GHz        265.0
96           Intel Atom x5-Z8350 1.44GHz        255.0
97           Intel Atom X5-Z8350 1.44GHz        210.8
98  Intel Celeron Dual Core 3205U 1.5GHz        209.0

[99 rows x 2 columns]

In [30]:
#ADEMAS LE ESTOY AÑADIENDO LA COLUMNA CPU_NORM QUE ES A LA QUE LE VOY A DAR EL VALOR QUE VA A IR DESDE 0.98 A 0 (POR QUÉ ESTE VALR, PORQUE HAY 99 FILAS EN EL DATAFRAME AGRUPADO, ES DECIR, 99 TIPOS DISNTINTOS DECPU), ASIGNANDO EL MAYOR VALOR AL PRECIO MAS CARO.
df_grou_cpu['Cpu_norm'] =0
df_grou_cpu

Cpu  Price_euros  Cpu_norm
0            Intel Core i7 7820HK 2.9GHz       6099.0         0
1          Intel Xeon E3-1535M v6 3.1GHz       4899.0         0
2          Intel Xeon E3-1535M v5 2.9GHz       4389.0         0
3            Intel Core i7 6820HK 2.7GHz       3975.0         0
4            Intel Core i7 6700HQ 2.6GHz       3949.4         0
..                                   ...          ...       ...
94              AMD E-Series 7110 1.8GHz        297.0         0
95              Intel Atom Z8350 1.92GHz        265.0         0
96           Intel Atom x5-Z8350 1.44GHz        255.0         0
97           Intel Atom X5-Z8350 1.44GHz        210.8         0
98  Intel Celeron Dual Core 3205U 1.5GHz        209.0         0

[99 rows x 3 columns]

In [31]:
# AQUI ESTOY ITRENANDO A TRAVES DEL DATAFRAME CREADO PARA AÑADIR EL VALOR CORREPONDEINETE A LA COLUMNA GPU_NORM
valor = 0.98
for i in range((df_grou_cpu.shape[0])):
    df_grou_cpu.iat[i,df_grou_cpu.columns.get_loc('Cpu_norm')]=valor
    valor -= 0.01

In [32]:

# DESPUES PASO A DICCIONARIO LA COLUMNA CPU Y CPU_NORM PARA DESPUES REEMPLAZARLO EN TRAIN
dic_Cpu = pd.Series(df_grou_cpu.Cpu_norm.values, index=df_grou_cpu.Cpu).to_dict()

In [33]:
dic_Cpu


{'Intel Core i7 7820HK 2.9GHz': 0.98,
 'Intel Xeon E3-1535M v6 3.1GHz': 0.97,
 'Intel Xeon E3-1535M v5 2.9GHz': 0.96,
 'Intel Core i7 6820HK 2.7GHz': 0.95,
 'Intel Core i7 6700HQ 2.6GHz': 0.94,
 'Intel Core i7 7700HQ 2.8GHz': 0.9299999999999999,
 'Intel Core i7 7820HQ 2.9GHz': 0.9199999999999999,
 'Intel Core i7 2.9GHz': 0.9099999999999999,
 'Intel Core i7 7500U 2.7GHz': 0.8999999999999999,
 'Intel Core i7 6600U 2.6GHz': 0.8899999999999999,
 'Intel Core i7 7660U 2.5GHz': 0.8799999999999999,
 'Intel Core i7 7600U 2.8GHz': 0.8699999999999999,
 'Intel Core i7 2.7GHz': 0.8599999999999999,
 'Intel Core i7 8550U 1.8GHz': 0.8499999999999999,
 'Intel Core i7 2.8GHz': 0.8399999999999999,
 'Intel Core i7 6820HQ 2.7GHz': 0.8299999999999998,
 'Intel Core i7 6500U 2.5GHz': 0.8199999999999998,
 'Intel Core i5 7200U 2.5GHz': 0.8099999999999998,
 'Intel Core i7 6500U 2.50GHz': 0.7999999999999998,
 'Intel Core i5 6300HQ 2.3GHz': 0.7899999999999998,
 'AMD Ryzen 1700 3GHz': 0.7799999999999998,
 'Intel Co

In [34]:
#AQUI HAGO EL REPLACE DE LOS VALORES DEL DICCIONARIO EN TRAIN
train['Cpu']= train['Cpu'].replace(dic_Cpu)

In [35]:

# AQUI ESTOY AGRUPANDO POR LA COLUMNA CPU, Y MOSTRANDO EL PRECIO MAXIMO POR CPU ORDENADOR DE FORMA DESCENDENTE Y SETEO EL INDICE DEL NUEVO DATAFRAME CREADO PARA PODER DESPUES ITERAR A TRAVES DE LOS INDICES


df_grou_ScreenResolution = train.groupby(['ScreenResolution'])[['Price_euros']].max().sort_values(by='Price_euros',ascending=False)
df_grou_ScreenResolution.reset_index(inplace=True)
df_grou_ScreenResolution



ScreenResolution  Price_euros
0             4K Ultra HD / Touchscreen 3840x2160      6099.00
1                 IPS Panel 4K Ultra HD 3840x2160      4899.00
2                     IPS Panel Full HD 1920x1080      4389.00
3                           4K Ultra HD 3840x2160      3659.40
4                               Full HD 1920x1080      3499.00
5              IPS Panel Retina Display 2880x1800      2858.00
6                           Touchscreen 2560x1440      2824.00
7                             IPS Panel 2560x1440      2799.00
8                                       2560x1440      2758.00
9                Quad HD+ / Touchscreen 3200x1800      2680.00
10                          Touchscreen 2256x1504      2589.00
11                Full HD / Touchscreen 1920x1080      2559.00
12  IPS Panel 4K Ultra HD / Touchscreen 3840x2160      2449.00
13                IPS Panel Touchscreen 2560x1440      2440.00
14  IPS Panel Touchscreen / 4K Ultra HD 3840x2160      2250.68
15      IPS Panel Full HD / Touchscreen 1920x1080      2099.00
16                   IPS Panel Quad HD+ 2560x1440      2099.00
17             IPS Panel Retina Display 2560x1600      1958.90
18                    IPS Panel Full HD 2560x1440      1943.00
19                                       1366x768      1895.00
20               Touchscreen / Quad HD+ 3200x1800      1869.00
21                                       1600x900      1860.99
22                          Touchscreen 2400x1600      1559.00
23             IPS Panel Retina Display 2304x1440      1510.00
24     IPS Panel Quad HD+ / Touchscreen 3200x1800      1499.00
25                    IPS Panel Full HD 2160x1440      1499.00
26                Touchscreen / Full HD 1920x1080      1358.00
27                             Quad HD+ 3200x1800      1268.00
28                                       1440x900      1158.70
29                   IPS Panel Quad HD+ 3200x1800      1145.00
30                     IPS Panel Full HD 1366x768       902.00
31                           Touchscreen 1366x768       739.00
32                             IPS Panel 1366x768       722.00
33                IPS Panel Touchscreen 1920x1200       646.27
34                                      1920x1080       581.90
35                 IPS Panel Touchscreen 1366x768       379.00
36                    IPS Panel Full HD 1920x1200       265.00

In [36]:
df_grou_ScreenResolution.shape

(37, 2)

In [37]:
#ADEMAS LE ESTOY AÑADIENDO LA COLUMNA ScreenResolution_NORM QUE ES A LA QUE LE VOY A DAR EL VALOR QUE VA A IR DESDE 0.36 A 0 (POR QUÉ ESTE VALR, PORQUE HAY 0.36 FILAS EN EL DATAFRAME AGRUPADO, ES DECIR, 37 TIPOS DISNTINTOS DECPU), ASIGNANDO EL MAYOR VALOR AL PRECIO MAS CARO.
df_grou_ScreenResolution['ScreenResolution_norm'] =0
df_grou_ScreenResolution

ScreenResolution  Price_euros  \
0             4K Ultra HD / Touchscreen 3840x2160      6099.00   
1                 IPS Panel 4K Ultra HD 3840x2160      4899.00   
2                     IPS Panel Full HD 1920x1080      4389.00   
3                           4K Ultra HD 3840x2160      3659.40   
4                               Full HD 1920x1080      3499.00   
5              IPS Panel Retina Display 2880x1800      2858.00   
6                           Touchscreen 2560x1440      2824.00   
7                             IPS Panel 2560x1440      2799.00   
8                                       2560x1440      2758.00   
9                Quad HD+ / Touchscreen 3200x1800      2680.00   
10                          Touchscreen 2256x1504      2589.00   
11                Full HD / Touchscreen 1920x1080      2559.00   
12  IPS Panel 4K Ultra HD / Touchscreen 3840x2160      2449.00   
13                IPS Panel Touchscreen 2560x1440      2440.00   
14  IPS Panel Touchscreen / 4K Ultra HD 3840x2160      2250.68   
15      IPS Panel Full HD / Touchscreen 1920x1080      2099.00   
16                   IPS Panel Quad HD+ 2560x1440      2099.00   
17             IPS Panel Retina Display 2560x1600      1958.90   
18                    IPS Panel Full HD 2560x1440      1943.00   
19                                       1366x768      1895.00   
20               Touchscreen / Quad HD+ 3200x1800      1869.00   
21                                       1600x900      1860.99   
22                          Touchscreen 2400x1600      1559.00   
23             IPS Panel Retina Display 2304x1440      1510.00   
24     IPS Panel Quad HD+ / Touchscreen 3200x1800      1499.00   
25                    IPS Panel Full HD 2160x1440      1499.00   
26                Touchscreen / Full HD 1920x1080      1358.00   
27                             Quad HD+ 3200x1800      1268.00   
28                                       1440x900      1158.70   
29                   IPS Panel Quad HD+ 3200x1800      1145.00   
30                     IPS Panel Full HD 1366x768       902.00   
31                           Touchscreen 1366x768       739.00   
32                             IPS Panel 1366x768       722.00   
33                IPS Panel Touchscreen 1920x1200       646.27   
34                                      1920x1080       581.90   
35                 IPS Panel Touchscreen 1366x768       379.00   
36                    IPS Panel Full HD 1920x1200       265.00   

    ScreenResolution_norm  
0                       0  
1                       0  
2                       0  
3                       0  
4                       0  
5                       0  
6                       0  
7                       0  
8                       0  
9                       0  
10                      0  
11                      0  
12                      0  
13                      0  
14                      0  
15                      0  
16                      0  
17                      0  
18                      0  
19                      0  
20                      0  
21                      0  
22                      0  
23                      0  
24                      0  
25                      0  
26                      0  
27                      0  
28                      0  
29                      0  
30                      0  
31                      0  
32                      0  
33                      0  
34                      0  
35                      0  
36                      0

In [38]:
# AQUI ESTOY ITRENANDO A TRAVES DEL DATAFRAME CREADO PARA AÑADIR EL VALOR CORREPONDEINETE A LA COLUMNA ScreenResolution_NOR

valor = 0.36
for i in range((df_grou_ScreenResolution.shape[0])):
    df_grou_ScreenResolution.iat[i,df_grou_ScreenResolution.columns.get_loc('ScreenResolution_norm')]=valor
    valor -= 0.01

In [39]:
df_grou_ScreenResolution

ScreenResolution  Price_euros  \
0             4K Ultra HD / Touchscreen 3840x2160      6099.00   
1                 IPS Panel 4K Ultra HD 3840x2160      4899.00   
2                     IPS Panel Full HD 1920x1080      4389.00   
3                           4K Ultra HD 3840x2160      3659.40   
4                               Full HD 1920x1080      3499.00   
5              IPS Panel Retina Display 2880x1800      2858.00   
6                           Touchscreen 2560x1440      2824.00   
7                             IPS Panel 2560x1440      2799.00   
8                                       2560x1440      2758.00   
9                Quad HD+ / Touchscreen 3200x1800      2680.00   
10                          Touchscreen 2256x1504      2589.00   
11                Full HD / Touchscreen 1920x1080      2559.00   
12  IPS Panel 4K Ultra HD / Touchscreen 3840x2160      2449.00   
13                IPS Panel Touchscreen 2560x1440      2440.00   
14  IPS Panel Touchscreen / 4K Ultra HD 3840x2160      2250.68   
15      IPS Panel Full HD / Touchscreen 1920x1080      2099.00   
16                   IPS Panel Quad HD+ 2560x1440      2099.00   
17             IPS Panel Retina Display 2560x1600      1958.90   
18                    IPS Panel Full HD 2560x1440      1943.00   
19                                       1366x768      1895.00   
20               Touchscreen / Quad HD+ 3200x1800      1869.00   
21                                       1600x900      1860.99   
22                          Touchscreen 2400x1600      1559.00   
23             IPS Panel Retina Display 2304x1440      1510.00   
24     IPS Panel Quad HD+ / Touchscreen 3200x1800      1499.00   
25                    IPS Panel Full HD 2160x1440      1499.00   
26                Touchscreen / Full HD 1920x1080      1358.00   
27                             Quad HD+ 3200x1800      1268.00   
28                                       1440x900      1158.70   
29                   IPS Panel Quad HD+ 3200x1800      1145.00   
30                     IPS Panel Full HD 1366x768       902.00   
31                           Touchscreen 1366x768       739.00   
32                             IPS Panel 1366x768       722.00   
33                IPS Panel Touchscreen 1920x1200       646.27   
34                                      1920x1080       581.90   
35                 IPS Panel Touchscreen 1366x768       379.00   
36                    IPS Panel Full HD 1920x1200       265.00   

    ScreenResolution_norm  
0            3.600000e-01  
1            3.500000e-01  
2            3.400000e-01  
3            3.300000e-01  
4            3.200000e-01  
5            3.100000e-01  
6            3.000000e-01  
7            2.900000e-01  
8            2.800000e-01  
9            2.700000e-01  
10           2.600000e-01  
11           2.500000e-01  
12           2.400000e-01  
13           2.300000e-01  
14           2.200000e-01  
15           2.100000e-01  
16           2.000000e-01  
17           1.900000e-01  
18           1.800000e-01  
19           1.700000e-01  
20           1.600000e-01  
21           1.500000e-01  
22           1.400000e-01  
23           1.300000e-01  
24           1.200000e-01  
25           1.100000e-01  
26           1.000000e-01  
27           9.000000e-02  
28           8.000000e-02  
29           7.000000e-02  
30           6.000000e-02  
31           5.000000e-02  
32           4.000000e-02  
33           3.000000e-02  
34           2.000000e-02  
35           1.000000e-02  
36          -1.977585e-16

In [40]:
# DESPUES PASO A DICCIONARIO LA COLUMNA ScreenResolution Y ScreenResolution_norm PARA DESPUES REEMPLAZARLO EN TRAIN
dic_ScreenResolution = pd.Series(df_grou_ScreenResolution.ScreenResolution_norm.values, index=df_grou_ScreenResolution.ScreenResolution).to_dict()

In [41]:
#AQUI HAGO EL REPLACE DE LOS VALORES DEL DICCIONARIO EN TRAIN
train['ScreenResolution']= train['ScreenResolution'].replace(dic_ScreenResolution)

In [42]:
train[(train['Product']== 'Pixelbook (Core')&(train['Weight']=='1.1kg')]

id  Company          Product  TypeName  Inches  ScreenResolution   Cpu  \
576  472        7  Pixelbook (Core         3    12.3              0.14  0.62   
600  762        7  Pixelbook (Core         3    12.3              0.14  0.62   

     Ram     Memory   Gpu  OpSys Weight  Price_euros  
576    4  128GB SSD  0.72      3  1.1kg       1275.0  
600    4  256GB SSD  0.72      3  1.1kg       1559.0

In [43]:
train.corr()

id   Company  TypeName    Inches  ScreenResolution  \
id                1.000000  0.033056 -0.022308 -0.087431         -0.116985   
Company           0.033056  1.000000 -0.040361  0.034321          0.049007   
TypeName         -0.022308 -0.040361  1.000000  0.690278          0.292976   
Inches           -0.087431  0.034321  0.690278  1.000000          0.248100   
ScreenResolution -0.116985  0.049007  0.292976  0.248100          1.000000   
Cpu              -0.027203  0.149013  0.073893  0.204337          0.459874   
Ram              -0.055113  0.050178  0.078431  0.224249          0.418498   
Gpu               0.026148  0.107980 -0.031763  0.000330          0.310235   
OpSys             0.039199  0.035736 -0.011934 -0.000434          0.141063   
Price_euros       0.069513  0.108591 -0.059046  0.085601          0.385603   

                       Cpu       Ram       Gpu     OpSys  Price_euros  
id               -0.027203 -0.055113  0.026148  0.039199     0.069513  
Company           0.149013  0.050178  0.107980  0.035736     0.108591  
TypeName          0.073893  0.078431 -0.031763 -0.011934    -0.059046  
Inches            0.204337  0.224249  0.000330 -0.000434     0.085601  
ScreenResolution  0.459874  0.418498  0.310235  0.141063     0.385603  
Cpu               1.000000  0.666563  0.639340  0.183874     0.621806  
Ram               0.666563  1.000000  0.434421  0.177337     0.740558  
Gpu               0.639340  0.434421  1.000000  0.204101     0.548880  
OpSys             0.183874  0.177337  0.204101  1.000000     0.236838  
Price_euros       0.621806  0.740558  0.548880  0.236838     1.000000

In [44]:
train.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 912 entries, 0 to 911
Data columns (total 13 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   id                912 non-null    int64  
 1   Company           912 non-null    int64  
 2   Product           912 non-null    object 
 3   TypeName          912 non-null    int64  
 4   Inches            912 non-null    float64
 5   ScreenResolution  912 non-null    float64
 6   Cpu               912 non-null    float64
 7   Ram               912 non-null    int64  
 8   Memory            912 non-null    object 
 9   Gpu               912 non-null    float64
 10  OpSys             912 non-null    int64  
 11  Weight            912 non-null    object 
 12  Price_euros       912 non-null    float64
dtypes: float64(5), int64(5), object(3)
memory usage: 92.8+ KB


In [45]:
#sns.pairplot(train)

In [46]:
# plt.figure(figsize=(10,5))

# sns.heatmap(train.corr(),
#            vmin=-0.1,
#            vmax=1,
#            cmap=sns.diverging_palette(145, 280, s=85, l=25, n=7),
#            annot=True);

In [47]:
#HACIENDO EL SLICE PARA HACER LAS PRUEBAS PARA CREAR LA NUEVA COLUMNA NORMALIZADA

# a = train.iloc[0:15 ]
# a.sort_values(by='Price_euros', ascending=False, inplace=True)
# a.reset_index(inplace=True, drop=True)
# a

In [48]:
#CREANDO LA COLUMNA NORMALIZADA EN EL SLICE A PARA HACER LAS PRUEBAS DE REEMPLAZO DE VALORES DE LA COLUMNA GPU
# a['Gpu_norm'] =''
# a

In [49]:
#CODIGO DE PRUEBA PARA INTERNTAR REEMPLAZAR TODOS LOS VALORES DE LA COLUMNA GPU. FUNCIONA

# valor= 1
# a['Gpu_norm'] =''

# for index in range(a.shape[0]):
#     if index == 0:

#         a.at[index,'Gpu_norm']=valor      
        
#     else:
        
#         if a.iloc[ : , 9:10].values[index][0] == a.iloc[ : , 9:10].values[index-1][0]:

#             a.at[index,'Gpu_norm']=valor
#         else:
#             valor -= 0.01

#             a.at[index,'Gpu_norm']=valor
            
#a

In [50]:
# train.sort_index
# train

In [51]:
# UNA VEZ PROBADO EL CODIGO ANTERIOR PROCEDO A ORDENAR EL DATA TRAIN, RESETEO EL INDICE Y LE AÑADO LA COLUMNA DE Gpu_norm

# train.sort_values(by=['Gpu'],ascending=False, inplace=True)
# train.reset_index(inplace=True, drop=True)
# train['Gpu_norm'] =''

-----------------------------------------------------------------------------------------------------------------

## Pasos que si o si debéis realizar para poder participar en la competición:

### 1. Definir X e y

In [52]:
X = train.drop(['Price_euros'], axis=1)
y = train['Price_euros'].copy()
X.shape

(912, 12)

In [53]:
y.shape

(912,)

### 2. Dividir X_train, X_test, y_train, y_test

In [54]:
from sklearn.model_selection import train_test_split

In [55]:
X_train, X_test, y_train, y_test = train_test_split(X[['Cpu','Ram']], y, test_size = 0.20, random_state = 42)

In [56]:
X_train

Cpu  Ram
25   0.81    4
84   0.81    2
10   0.43    2
342  0.90    6
890  0.02    2
..    ...  ...
106  0.81    4
270  0.67    6
860  0.93    6
435  0.81    4
102  0.81    3

[729 rows x 2 columns]

### 3. Asignar el modelo (vacío) a una variable


In [57]:
#VOY A ENTRENAR AHORA UN RADONTREEREGRESION

In [58]:
from sklearn.ensemble import RandomForestRegressor

tree_reg = RandomForestRegressor(max_depth=3, random_state=42, n_estimators=150)
tree_reg.fit(X_train, y_train)

RandomForestRegressor(max_depth=3, n_estimators=150, random_state=42)

### 4. Sacar métricas, valorar el modelo 

Recuerda que en la competición se va a evaluar con la métrica de MAE.

In [59]:
prediciones_tre_reg_train = tree_reg.predict(X_train)
prediciones_tre_reg_test = tree_reg.predict(X_test)

In [60]:
from sklearn.metrics import mean_absolute_error

In [61]:
from sklearn.metrics import mean_squared_error

In [62]:

#4. SACO LAS METRICAS DE ERROR ENTRE LOS TRAIN Y LOS TEST PARA VER COMO HA CAMBIADO
print('Train MAE:', mean_absolute_error(y_train, prediciones_tre_reg_train))
print('Test MAE:', mean_absolute_error(y_test, prediciones_tre_reg_test),'\n')

Train MAE: 286.5751961936092
Test MAE: 279.46149810708044 



-----------------------------------------------------------------

## Una vez listo el modelo, toca predecir con el dataset de predicción 

Definición de **modelo que está listo**. 

Tras hacer suficientes pruebas, analizar los datos, hacer feature engineering, probar diferentes modelos con diferentes parámetros, es con este con el que observo mejores métricas y menos overfitting. ¡Cuidado con el overfitting aquí! Si vuestro modelo aprende muy bien de estos datos pero hay overfitting cuando le pasemos los datos desconocidos de `test.csv` nos arriesgamos a que digamos, no salga lo esperado.

### 1. Entrena dicho modelo con TODOS tus datos de train, esto es con `train.csv` al completo.


**CON LAS TRANSFORMACIONES QUE LE HAYAS REALIZADO A `X` INCLUÍDAS.**


Véase:
- Estandarización/Normalización
- Eliminación de Outliers
- Eliminación de columnas
- Creación de columnas nuevas
- Gestión de valores nulos
- Y un largo etcétera de técnicas que como Data Scientist hayas considerado las mejores para tu dataset.

### 2. Carga los datos de `test.csv` para predecir.

In [63]:
X_pred = pd.read_csv("test.csv")
X_pred.head()

id Company              Product   TypeName  Inches  \
0  1184      HP  Pavilion 15-BC000nv   Notebook    15.6   
1   815  Lenovo       Thinkpad T470s  Ultrabook    14.0   
2  1244      HP        EliteBook 840   Notebook    14.0   
3  1121      HP          Spectre Pro   Notebook    13.3   
4   443  Lenovo    IdeaPad 320-15IAP   Notebook    15.6   

              ScreenResolution                                   Cpu  Ram  \
0  IPS Panel Full HD 1920x1080           Intel Core i7 6700HQ 2.6GHz  6GB   
1            Full HD 1920x1080            Intel Core i7 7500U 2.7GHz  8GB   
2            Full HD 1920x1080            Intel Core i5 6200U 2.3GHz  4GB   
3            Full HD 1920x1080            Intel Core i5 6200U 2.3GHz  8GB   
4            Full HD 1920x1080  Intel Pentium Quad Core N4200 1.1GHz  4GB   

      Memory                      Gpu       OpSys  Weight  
0    1TB HDD  Nvidia GeForce GTX 960M  Windows 10  2.18kg  
1  256GB SSD    Intel HD Graphics 620  Windows 10  1.32kg  
2  256GB SSD    Intel HD Graphics 520   Windows 7  1.54kg  
3  256GB SSD    Intel HD Graphics 520  Windows 10  1.16kg  
4  128GB SSD    Intel HD Graphics 505  Windows 10   2.2kg

In [64]:
X_pred.shape

(391, 12)

In [65]:
# POR EL MOMENTO, LAS TRANSFORMACIONES QUE VOY A APLICAR VAN A SER SOBRE LAS COLUMNAS QUE UTILIZO EN EL ENTRANAMIENTO:

# SCREENRESOLUTION, CPU, RAM, GPU


In [66]:
# ESTE DICCIONARIO ES PARA CAMBIAR LOS VALORES DE LA RAM

dic_ram = {'8GB' : 4, '6GB': 3, '4GB' : 2, '16GB' : 6, '24GB' : 7, '12GB' : 5, '64GB': 9, '32GB': 8, '2GB' :1}
X_pred.groupby(['Ram']).max().sort_index()

id  Company              Product     TypeName  Inches  \
Ram                                                             
12GB   962       HP          ZenBook Pro     Notebook    17.3   
16GB  1271  Toshiba            ZenBook 3  Workstation    17.3   
24GB   517     Asus         ROG Zephyrus       Gaming    15.6   
2GB   1264   Lenovo        VivoBook Flip     Notebook    15.6   
32GB   901  Toshiba               XPS 15    Ultrabook    18.4   
4GB   1298  Toshiba            Yoga Book    Ultrabook    17.3   
6GB   1276   Lenovo  Pavilion 15-BC000nv     Notebook    17.3   
8GB   1293   Xiaomi            Zenbook 3  Workstation    17.3   

                 ScreenResolution                                   Cpu  \
Ram                                                                       
12GB  IPS Panel Full HD 1920x1080           Intel Core i7 7700HQ 2.8GHz   
16GB        Touchscreen 2560x1440           Intel Xeon E3-1505M V6 3GHz   
24GB            Full HD 1920x1080           Intel Core i7 7700HQ 2.8GHz   
2GB          Touchscreen 1366x768  Intel Celeron Dual Core N3350 1.1GHz   
32GB  IPS Panel Full HD 1920x1080           Intel Core i7 7820HK 2.9GHz   
4GB          Touchscreen 1366x768         Samsung Cortex A72&A53 2.0GHz   
6GB          Touchscreen 1366x768            Intel Core i7 7500U 2.7GHz   
8GB         Touchscreen 2256x1504            Intel Core i7 8650U 1.9GHz   

                      Memory                      Gpu       OpSys  Weight  
Ram                                                                        
12GB                 1TB HDD  Nvidia GeForce GTX 950M  Windows 10  3.74kg  
16GB    512GB SSD +  1TB HDD       Nvidia Quadro M620   Windows 7   4.6kg  
24GB               512GB SSD   Nvidia GeForce GTX1080  Windows 10  2.24kg  
2GB                500GB HDD    Intel HD Graphics 500  Windows 10  2.20kg  
32GB    512GB SSD +  1TB HDD  Nvidia GeForce GTX 980M  Windows 10   4.7kg  
4GB                 64GB SSD     Nvidia GeForce MX150   Windows 7     3kg  
6GB       64GB Flash Storage  Nvidia GeForce GTX 960M  Windows 10   2.8kg  
8GB   512GB SSD +  512GB SSD      Nvidia Quadro M1200       macOS  4.42kg

In [67]:

# CAMBIO LOS VALORES DE RAM POR LOS DEL DICCIONARIO

X_pred ['Ram'] =X_pred['Ram'].replace(dic_ram)


In [68]:
# # AQUI ESTOY AGRUPANDO POR LA COLUMNA GPU, Y MOSTRANDO EL PRECIO MAXIMO POR GPU ORDENADOR DE FORMA DESCENDENTE Y SETEO EL INDICE DEL NUEVO DATAFRAME CREADO PARA PODER DESPUES ITERAR A TRAVES DE LOS INDICES


# df_grou_gpu = X_pred.groupby(['Gpu']).max().sort_index()
# df_grou_gpu.reset_index(inplace=True)
# #ADEMAS LE ESTOY AÑADIENDO LA COLUMNA GPU_NORM QUE ES A LA QUE LE VOY A DAR EL VALOR QUE VA A IR DESDE 1 A 0 (POR QUÉ ESTE VALR, PORQUE HAY 100 FILAS EN EL DATAFRAME AGRUPADO, ES DECIR, 100 TIPOS DISNTINTOS DECPU), ASIGNANDO EL MAYOR VALOR AL PRECIO MAS CARO.
# df_grou_gpu['Gpu_norm'] =0
# df_grou_gpu

In [69]:
# AQUI REEMPLAZAMOS YA LA COLUMNA DE GPU CON EL PRIMER DICCIONARIO QUE HEMOS CREADO PARA ESTA COLUMNA
X_pred['Gpu']= X_pred['Gpu'].replace(dic_gpu)

In [70]:
X_pred['Gpu'].unique()

array([0.5799999999999996, 0.8699999999999999, 0.8599999999999999,
       0.22999999999999932, 0.28999999999999937, 0.6499999999999997,
       0.7999999999999998, 0.5199999999999996, 0.09999999999999924,
       0.8899999999999999, 0.8299999999999998, 0.6999999999999997,
       0.5599999999999996, 0.6299999999999997, 0.12999999999999923, 0.99,
       0.15999999999999925, 0.9199999999999999, 0.3199999999999994,
       0.4199999999999995, 'AMD FirePro W5130M', 0.7299999999999998, 0.94,
       0.6599999999999997, 0.17999999999999927, 0.7199999999999998,
       0.5899999999999996, 0.24999999999999933, 0.009999999999999247,
       0.5299999999999996, 'AMD Radeon R7 M465', 0.11999999999999923,
       0.18999999999999928, 0.02999999999999925, 0.06999999999999926,
       0.36999999999999944, 0.3399999999999994, 'ARM Mali T860 MP4',
       0.13999999999999924, 0.48999999999999955, 0.2099999999999993,
       0.7699999999999998, 0.2199999999999993, 0.6199999999999997,
       0.6899999999999997, 'N

In [71]:
dic_gpu_x_pred = {'AMD FirePro W5130M': 0.5 , 'AMD Radeon R7 M465':0.5, 'ARM Mali T860 MP4':0.5, 'Nvidia GeForce GTX1080':0.5,  'AMD Radeon R5 520':0.5, 'Intel Iris Plus Graphics 650':0.5,'AMD FirePro W6150M':0.5, -7.528699885739343e-16:0.5,
                  'Nvidia Quadro 3000M':0.5,'Intel HD Graphics 530':0.5,'Nvidia GTX 980 SLI':0.5}

X_pred['Gpu']= X_pred['Gpu'].replace(dic_gpu_x_pred)

In [72]:
X_pred.describe()

id      Inches         Ram         Gpu
count   391.000000  391.000000  391.000000  391.000000
mean    646.156010   15.049616    3.639386    0.663427
std     370.544552    1.402303    1.493708    0.271592
min       4.000000   10.100000    1.000000    0.010000
25%     350.500000   14.000000    2.000000    0.500000
50%     635.000000   15.600000    4.000000    0.800000
75%     960.500000   15.600000    4.000000    0.870000
max    1298.000000   18.400000    8.000000    0.990000

In [73]:
X_pred

id Company              Product   TypeName  Inches  \
0    1184      HP  Pavilion 15-BC000nv   Notebook    15.6   
1     815  Lenovo       Thinkpad T470s  Ultrabook    14.0   
2    1244      HP        EliteBook 840   Notebook    14.0   
3    1121      HP          Spectre Pro   Notebook    13.3   
4     443  Lenovo    IdeaPad 320-15IAP   Notebook    15.6   
..    ...     ...                  ...        ...     ...   
386   739    Acer       Aspire ES1-572   Notebook    15.6   
387   360      HP               250 G6   Notebook    15.6   
388   965    Dell        Inspiron 3567   Notebook    15.6   
389    82      HP          ProBook 440   Notebook    14.0   
390  1276    Dell        Inspiron 3567   Notebook    15.6   

                ScreenResolution                                   Cpu  Ram  \
0    IPS Panel Full HD 1920x1080           Intel Core i7 6700HQ 2.6GHz    3   
1              Full HD 1920x1080            Intel Core i7 7500U 2.7GHz    4   
2              Full HD 1920x1080            Intel Core i5 6200U 2.3GHz    2   
3              Full HD 1920x1080            Intel Core i5 6200U 2.3GHz    4   
4              Full HD 1920x1080  Intel Pentium Quad Core N4200 1.1GHz    2   
..                           ...                                   ...  ...   
386                     1366x768              Intel Core i3 6006U 2GHz    2   
387                     1366x768              Intel Core i3 6006U 2GHz    2   
388                     1366x768            Intel Core i3 6006U 2.0GHz    2   
389            Full HD 1920x1080            Intel Core i5 8250U 1.6GHz    4   
390                     1366x768            Intel Core i3 7100U 2.4GHz    3   

        Memory   Gpu       OpSys  Weight  
0      1TB HDD  0.58  Windows 10  2.18kg  
1    256GB SSD  0.87  Windows 10  1.32kg  
2    256GB SSD  0.86   Windows 7  1.54kg  
3    256GB SSD  0.86  Windows 10  1.16kg  
4    128GB SSD  0.23  Windows 10   2.2kg  
..         ...   ...         ...     ...  
386  500GB HDD  0.86  Windows 10   2.4kg  
387  500GB HDD  0.86  Windows 10  1.86kg  
388    1TB HDD  0.29       Linux  2.25kg  
389  256GB SSD  0.87  Windows 10  1.63kg  
390    1TB HDD  0.87  Windows 10   2.3kg  

[391 rows x 12 columns]

In [74]:
#  VAMOS AHORA A CAMBIAR LOS VALORES DE LA COLUMNA CPU, ASI QUE VAMOS A UTILIZAR PRIMERO EL DICCIONARIO QUE HEMOS CREADO Y LUEGO COMPROBAMOS LOS VALORES QUE NO SE HAN PODIDO REEMPLAZAR PARA VER LOS VALORES NUEVOS Y CREAR UN NUEVO DICC PARA REEMPLAZARLO
#AQUI HAGO EL REPLACE DE LOS VALORES DEL DICCIONARIO EN TRAIN
X_pred['Cpu']= X_pred['Cpu'].replace(dic_Cpu)

In [75]:
X_pred['Cpu'].unique()

array([0.94, 0.8999999999999999, 0.6399999999999997, 0.39999999999999947,
       0.6999999999999997, 0.8099999999999998, 0.9299999999999999,
       0.7599999999999998, 0.8499999999999999, 0.24999999999999933,
       0.3299999999999994, 0.7899999999999998, 0.7099999999999997,
       0.4199999999999995, 0.15999999999999925, 0.98, 0.2099999999999993,
       0.5199999999999996, 0.3499999999999994, 0.5099999999999996,
       'Intel Core i5 6440HQ 2.6GHz',
       'Intel Pentium Dual Core 4405Y 1.5GHz', 0.8199999999999998,
       0.4299999999999995, 0.8699999999999999,
       'AMD A10-Series 9620P 2.5GHz', 0.7399999999999998,
       'AMD E-Series E2-6110 1.5GHz', 'AMD A9-Series A9-9420 3GHz',
       'Intel Core i7 6560U 2.2GHz', 0.6699999999999997,
       0.23999999999999932, 0.17999999999999927, 0.18999999999999928,
       'Intel Core i3 6100U 2.1GHz', 'Samsung Cortex A72&A53 2.0GHz',
       'Intel Celeron Dual Core N3060 1.60GHz', 0.26999999999999935,
       0.7299999999999998, 0.4699999999

In [76]:
#CREO EL DICCIONARIO CON LAS NUEVAS VARIABLES Y LE ASIGNO EL VALOR DE 0.5

dic_cpu_x_pred = {'Intel Core i5 6440HQ 2.6GHz': 0.5 , 'Intel Pentium Dual Core 4405Y 1.5GHz':0.5, 'AMD A10-Series 9620P 2.5GHz':0.5, 'AMD E-Series E2-6110 1.5GHz':0.5, 'AMD A9-Series A9-9420 3GHz' :0.5, 'Intel Core i7 6560U 2.2GHz':0.5,'Intel Core i3 6100U 2.1GHz':0.5,
                  'Samsung Cortex A72&A53 2.0GHz':0.5,'Intel Celeron Dual Core N3060 1.60GHz':0.5, 'Intel Core M m7-6Y75 1.2GHz' :0.5,'Intel Core i7 7700HQ 2.7GHz':0.5, 'Intel Core i5 3.1GHz':0.5,'AMD E-Series E2-9000 2.2GHz' :0.5, 'AMD E-Series 6110 1.5GHz' :0.5,
                  'Intel Core M m3 1.2GHz':0.5,'Intel Pentium Dual Core 4405U 2.1GHz' : 0.5, 'Intel Core i3 6006U 2.2GHz' :0.5,'Intel Atom x5-Z8300 1.44GHz' :0.5,'Intel Core i7 6920HQ 2.9GHz' :0.5}

X_pred['Cpu']= X_pred['Cpu'].replace(dic_cpu_x_pred)

In [77]:
X_pred.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 391 entries, 0 to 390
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   id                391 non-null    int64  
 1   Company           391 non-null    object 
 2   Product           391 non-null    object 
 3   TypeName          391 non-null    object 
 4   Inches            391 non-null    float64
 5   ScreenResolution  391 non-null    object 
 6   Cpu               391 non-null    float64
 7   Ram               391 non-null    int64  
 8   Memory            391 non-null    object 
 9   Gpu               391 non-null    float64
 10  OpSys             391 non-null    object 
 11  Weight            391 non-null    object 
dtypes: float64(3), int64(2), object(7)
memory usage: 36.8+ KB


In [78]:
# VAMOS CON LA COLUMNA SCREENRESOLUTION


#AQUI HAGO EL REPLACE DE LOS VALORES DEL DICCIONARIO EN TRAIN
X_pred['ScreenResolution']= X_pred['ScreenResolution'].replace(dic_ScreenResolution)

In [79]:
X_pred['ScreenResolution'].unique()

array([0.33999999999999997, 0.31999999999999995, 0.16999999999999982,
       0.2499999999999999, 0.029999999999999805, 0.07999999999999981,
       0.0899999999999998, 0.11999999999999979, 0.35, 0.04999999999999981,
       'Touchscreen / 4K Ultra HD 3840x2160', 0.1499999999999998,
       0.20999999999999985, 0.03999999999999981,
       'IPS Panel Touchscreen 2400x1600', 0.32999999999999996,
       'IPS Panel Retina Display 2736x1824', 0.23999999999999988,
       0.009999999999999802, 0.36, 0.18999999999999984,
       0.10999999999999979, 0.1399999999999998, 0.2699999999999999,
       0.12999999999999978, 0.29999999999999993, 0.2599999999999999,
       0.2799999999999999], dtype=object)

In [80]:
#CREO EL DICCIONARIO CON LAS NUEVAS VARIABLES Y LE ASIGNO EL VALOR DE 0.16
dic_screen_x_pred = {'Touchscreen / 4K Ultra HD 3840x2160' : 0.16, 'IPS Panel Touchscreen 2400x1600' : 0.16, 'IPS Panel Retina Display 2736x1824': 0.16 }

X_pred['ScreenResolution']= X_pred['ScreenResolution'].replace(dic_screen_x_pred)

In [81]:
X_pred.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 391 entries, 0 to 390
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   id                391 non-null    int64  
 1   Company           391 non-null    object 
 2   Product           391 non-null    object 
 3   TypeName          391 non-null    object 
 4   Inches            391 non-null    float64
 5   ScreenResolution  391 non-null    float64
 6   Cpu               391 non-null    float64
 7   Ram               391 non-null    int64  
 8   Memory            391 non-null    object 
 9   Gpu               391 non-null    float64
 10  OpSys             391 non-null    object 
 11  Weight            391 non-null    object 
dtypes: float64(4), int64(2), object(6)
memory usage: 36.8+ KB


**¿Por qué da error?** 

In [82]:
# VOY A ENTRENAR UN MODELO DE DECISION REGREESOR TREE


# from sklearn.tree import DecisionTreeRegressor

# tree_reg = DecisionTreeRegressor(max_depth=2, random_state=0)
# tree_reg.fit(X_train, y_train)

In [83]:
predictions_submit = tree_reg.predict(X_pred[['Cpu','Ram']])
predictions_submit.shape



(391,)

In [84]:
predictions_submit.shape

(391,)

IMPORTANTE: APLICAR LO MISMO A ESTOS DATOS QUE HAYÁIS APLICADO A LOS DATOS DE ENTRENAMIENTO

- SI EL ARRAY CON EL QUE HICISTEIS `.fit()` ERA DE 4 COLUMNAS, PARA `.predict()` DEBEN SER LAS MISMAS
- SI AL ARRAY CON EL QUE HICISTEIS `.fit()` LO NORMALIZASTEIS, PARA `.predict()` DEBÉIS NORMALIZARLO
- TODO IGUAL SALVO BORRAR FILAS, EL NÚMERO DE ROWS SE DEBE MANTENER EN ESTE SET, PUES LA PREDICCIÓN DEBE TENER 391 FILAS, SI O SI

**Entonces, si al cargar los datos de train usé `index_col=0` para que utilizara la primera columna del conjunto de datos como índice, ¿tendré que hacerlo también para el conjunto `test.csv`?**

In [85]:
# ¿Qué opináis?
# Sí

![](https://www.tierraljelechu.com/web/img/wiki_up/1.996-SorpresaDto.-1-Red.jpg)

In [86]:
# X_pred = pd.read_csv("Dataset/test.csv", index_col=0)
# X_pred.head()

### 3. AHORA puedo hacer la predicción que será lo que subirás a Kaggle. 

**¿Qué es lo que subirás a Kaggle?**

In [87]:
# predictions_submit = model.predict(X_pred)
# predictions_submit

**¡PERO! Para subir a Kaggle la predicción, ésta tendrá que tener una forma específica y no valdrá otra.**

En este caso, la **MISMA** forma que `sample_submission.csv`. 

In [88]:
sample = pd.read_csv("sample_submission.csv")

In [89]:
sample.head()

id  Price_euros
0  1184         2158
1   815          511
2  1244         5290
3  1121          622
4   443         1140

In [90]:
sample.shape

(391, 2)

### 4. Mete tus predicciones en un dataframe. 

En este caso, la **MISMA** forma que `sample_submission.csv`. 

In [91]:
submission = pd.DataFrame({"id": sample['id'], "Price_euros": predictions_submit})

#submission = pd.DataFrame({"id": range(len(predictions_submit)), "Price_euros": predictions_submit})

In [92]:
submission.head()

id  Price_euros
0  1184  1029.705971
1   815  1285.520152
2  1244   871.055846
3  1121  1161.505569
4   443   528.024089

In [93]:
submission.shape

(391, 2)

### 5. Pásale el CHEQUEATOR para comprobar que efectivamente está listo para subir a Kaggle.

In [94]:
def chequeator(df_to_submit):
    """
    Esta función se asegura de que tu submission tenga la forma requerida por Kaggle.
    
    Si es así, se guardará el dataframe en un `csv` y estará listo para subir a Kaggle.
    
    Si no, LEE EL MENSAJE Y HAZLE CASO.
    
    Si aún no:
    - apaga tu ordenador, 
    - date una vuelta, 
    - enciendelo otra vez, 
    - abre este notebook y 
    - leelo todo de nuevo. 
    Todos nos merecemos una segunda oportunidad. También tú.
    """
    if df_to_submit.shape == sample.shape:
        if df_to_submit.columns.all() == sample.columns.all():
            if df_to_submit.id.all() == sample.id.all():
                print("You're ready to submit!")
                submission.to_csv("submission_13.csv", index = False) #muy importante el index = False
                urllib.request.urlretrieve("https://i.kym-cdn.com/photos/images/facebook/000/747/556/27a.jpg", "gfg.png")     
                img = Image.open("gfg.png")
                img.show()   
            else:
                print("Check the ids and try again")
        else:
            print("Check the names of the columns and try again")
    else:
        print("Check the number of rows and/or columns and try again")
        print("\nMensaje secreto de Clara: No me puedo creer que después de todo este notebook hayas hecho algún cambio en las filas de `diamonds_test.csv`. Lloro.")


In [96]:
chequeator(submission)

You're ready to submit!
